src/ruby/puppetserver-lib/puppet/server/compiler.rb

```diff
@@ -0,0 +1,160 @@
+        Hash[hash.collect do |key, value|
+            # Stolen and modified from params_to_ruby in handler.rb
+            newkey = key.to_s
+            # Java::ClojureLang::Keywords retain their leading colon when
+            # converted to a string
+            newkey.delete_prefix!(':')
+
+            if value.java_kind_of?(Java::ClojureLang::IPersistentMap)
+              [newkey, convert_java_args_to_ruby(value)]
+            elsif value.java_kind_of?(Java::JavaUtil::List)
+              [newkey, value.to_a]
+            else
+              [newkey, value]
+            end
+          end]
+      end
+
+
+      # @return Puppet::Node::Facts facts, Hash trusted_facts
+      def process_facts(request_data)
+        facts = extract_facts(request_data)
+        trusted_facts = extract_trusted_facts(request_data, facts)
+
+        return facts, trusted_facts
+      end
+
+      def extract_facts(request_data)
+        if request_data['facts'].nil?
+          if Puppet[:storeconfigs] == true
+            facts = get_facts_from_pdb(request_data['certname'], request_data['environment'])
+          else
+            raise(Puppet::Error, "PuppetDB not configured, please provide facts with your catalog request.")
+          end
+        else
+          facts_from_request = request_data['facts']
+
+          # Ensure request data has the proper keys, mirroring the structure
+          # of Facts#to_data_hash
+          facts_from_request['values'] ||= {}
+          facts_from_request['name'] ||= request_data['certname']
+
+          facts = Puppet::Node::Facts.from_data_hash(facts_from_request)
+        end
+
+        facts.sanitize
+
+        facts
+      end
+
+      def extract_trusted_facts(request_data, facts)
+        # Pull the trusted facts from the request, or attempt to extract them from
+        # the facts hash
+        trusted_facts = if request_data['trusted_facts']
+                          request_data['trusted_facts']['values']
+                        else
+                          fact_values = facts.to_data_hash['values']
+                          fact_values['trusted']
+                        end
+        # If no trusted facts could be found, ensure a hash is returned
+        trusted_facts ||= {}
+      end
+
+      def get_facts_from_pdb(nodename, environment)
+        pdb_terminus = Puppet::Node::Facts::Puppetdb.new
+        request = Puppet::Indirector::Request.new(pdb_terminus.class.name,
+                                                  :find,
+                                                  nodename,
+                                                  nil,
+                                                  :environment => environment)
+        facts = pdb_terminus.find(request)
+
+        # If no facts have been stored for the node, PDB will return nil
+        if facts.nil?
+          # Create an empty facts object
+          facts = Puppet::Node::Facts.new(request_data['certname'])
```

Comment:
there's no request_data in scope here

Reply:
Oh good catch. I'll test this branch (the case where you're asking PDB for facts for a node that doesn't exist) to make sure the rest of it works okay.

Reply:
Confirmed providing no facts and PDB also not having facts works. Though I would like to know whether that is good behavior. Should we allow catalogs to be compiled when no facts are available?

Reply:
I don't think that's a valid thing to do, though I'd defer to Sorenson on that kind of thing.

Reply:
Okay. In that case maybe this should raise. It seems... wrong to me to allow that.